### ISSUE TITLE
Add Support for NET Core 3.0

### ISSUE THREAD

**Issue by zipper97412** [redhat-developer/s2i-dotnetcore#271]
Hi,
I would like to know if it is possible to add support for .net Core 3.0
I can work on a PR but I don't know if a simple search and replace for 2.2 -> 3.0 would do the trick
Any advice?

**Comment by omajid** [comment]
Hi

Adding support for .NET Core 3.0 is definitely in our plans.

Unfortunately, it's not trivial to implement just yet: if you look at the existing Dockerfiles, you will see that we generally rely on the each version of .NET Core being available on RHEL or CentOS before we can include it here. If a version of .NET Core is not available on those platforms, we have no simple way of adding it in a production-ready state to this repo.

However, for initial implementation and testing, it might be possible to do what https://github.com/redhat-developer/s2i-dotnetcore/pull/236 did; we could use the binaries published by Microsoft for 3.0. That might give us a way to get started with using 3.0 before it's added to RHEL or CentOS.

**Comment by geoberle** [comment]
The ‚binary-idea‘ would be really great. Lots of people are already eager to give 3.0 a try and sadly the images from Microsoft are not Openshift compatible (running as root, no S2I, ...).

**Comment by omajid** [comment]
Fixed by https://github.com/redhat-developer/s2i-dotnetcore/pull/286

**Comment by azixaka** [comment]
Could you please tell when will this get merged into master? And when will the .net core 3.0 images become available in the RH Container Catalogue?

**Comment by omajid** [comment]
Hi @azixaka 

A couple of thoughts:

The main pre-requisite for everything is having .NET Core 3.0 publicly available via [source-build](https://github.com/dotnet/source-build). source-build allows us *nix folks to build .NET Core from sources and avoid prebuilt binaries as much as possible. source-build folks (that includes us, in some capacities) are working on getting 3.0 ready, but it's just not there yet. We should have something by the end of this week unless something unexpected happens.

Once 3.0 GA is ready in source-build, we get this merged in master and start our various Red Hat-internal release processes to get the RPM packages and these containers built. Unfortunately that will take a few days, maybe a week or two. What we try to do is to make sure that our release, when it goes out, works and is free of known bugs. We want to avoid releasing something that has bugs and then having to release a fix for it a few hours or few days later.

So I can't really give you a firm date, but I can tell you we are working on it and .NET Core 3.0 packages and containers are pretty much the number 1 priority right now for me and my team.

**Comment by azixaka** [comment]
You mentioned above that you are going to use the binaries published by Microsoft to make it available quicker, like "almost same day as Microsoft release claimed previously" ;-) and probably switch source-build later?

Come on guys, it's 2019 and corporation we work in wants to move faster than what OpenShift's pace now. We were evaluating .net core 3 since early previous and ready now. It would be great if you could start doing all this work as part of the preview/RC stage so it would simply be ready by the release date rather than initiating it post the release date. We want to see all this recent Microsoft/Linux/RedHad love compliments exchange in action.

**Comment by tmds** [comment]
Anything from Red Hat needs to be built from source.
.NET Core 3.0 still had changes in past week(s) and source-build needs to catch-up. When there is a 3.0 tag, we can build rpms, and then containers.

We (Red Hat and Microsoft) want to have our release dates closely together, and we'll look at how we can improve.

PS: there's a lot of action going on in the .NET Core space.

**Comment by RheaAyase** [comment]
@azixaka to add to what my colleagues have said, since this is not obvious to public, we **must** wait for source-build, and we **can not** help move it along more than we already do. So your complaints are kinda in the wrong place.

You're not alone in these frustrations, we do our best to keep the linux release dates identical to the Microsoft ones, or at worst one day late, but in this case, we are stuck here.

**Comment by azixaka** [comment]
Thank you for your response.

In order to make it happen quicker maybe you could escalate this by commenting on this issue - [https://github.com/dotnet/source-build/issues/1256](url), so folks at Microsoft would know it prevents the whole enterprise world that happens to be on OpenShift from using .net core 3.0?

**Comment by azixaka** [comment]
The issue is now closed and Microsoft tagged the 3.0 source-build, available here - https://github.com/dotnet/source-build/releases/tag/v3.0.100-SDK

All eyes on you now, RedHat.

**Comment by Hoffs** [comment]
11 days?

**Comment by azixaka** [comment]
Yeah. They were telling it's due to Microsoft not tagging it timely and suggesting complaints were kind of in a wrong place... Now that it's been out for two weeks they stay very quiet, not commenting on anything.

It was exactly the same story with 2.1, 2.2... Surely that whole preparation described above could have been started at the time when Microsoft was releasing preview builds, so by the time Microsoft tagged it there is an image in OpenShift. The whole idea of OpenShift is to make sure customers get whatever is on the market with a delay. ;-)

**Comment by superslowloris** [comment]
I don't need this anymore. It was faster to update my system from Openshift to AKS.

**Comment by ninjasource** [comment]
Same here for us. We are .net core 3.0 ready, just waiting for Red Hat now. Unfortunately we may have to move away from this model because of this uncertainty.

**Comment by tmds** [comment]
We'll let you know when images are available.
Honestly, no one likes this delay. Definitely not the engineers who did do preparatory work for this.
Getting to published rpms and containers involves more process and people and takes some time.
To get close to Microsoft release date, we need a tag up-front.

We'll share your feedback internally and with Microsoft.

**Comment by RheaAyase** [comment]
RHEL7 RPM's and Containers are now released. Official announcements will follow via the usual channels.

**Comment by azixaka** [comment]
What a news! Thank you very much!

**Comment by geoberle** [comment]
thank you for making this happen!
